src/Microsoft.AspNetCore.Razor.Language/DefaultRazorIntermediateNodeLoweringPhase.cs

```diff
@@ -206,195 +207,220 @@ public LoweringVisitor(DocumentIntermediateNode document, IntermediateNodeBuilde
+                            Directive = CSharpCodeParser.TagHelperPrefixDirectiveDescriptor,
+                            Source = BuildSourceSpanFromNode(node),
+                        };
+                    }
+                    else
+                    {
+                        directiveNode = new DirectiveIntermediateNode()
+                        {
+                            DirectiveName = CSharpCodeParser.TagHelperPrefixDirectiveDescriptor.Directive,
+                            Directive = CSharpCodeParser.TagHelperPrefixDirectiveDescriptor,
+                            Source = BuildSourceSpanFromNode(node),
+                        };
+                    }
 
-                for (var i = 0; i < chunkGenerator.Diagnostics.Count; i++)
-                {
-                    directiveNode.Diagnostics.Add(chunkGenerator.Diagnostics[i]);
-                }
+                    for (var i = 0; i < tagHelperPrefixChunkGenerator.Diagnostics.Count; i++)
+                    {
+                        directiveNode.Diagnostics.Add(tagHelperPrefixChunkGenerator.Diagnostics[i]);
+                    }
 
-                _builder.Push(directiveNode);
+                    _builder.Push(directiveNode);
 
-                _builder.Add(new DirectiveTokenIntermediateNode()
-                {
-                    Content = chunkGenerator.Prefix,
-                    DirectiveToken = CSharpCodeParser.TagHelperPrefixDirectiveDescriptor.Tokens.First(),
-                    Source = BuildSourceSpanFromNode(span),
-                });
+                    _builder.Add(new DirectiveTokenIntermediateNode()
+                    {
+                        Content = tagHelperPrefixChunkGenerator.Prefix,
+                        DirectiveToken = CSharpCodeParser.TagHelperPrefixDirectiveDescriptor.Tokens.First(),
+                        Source = BuildSourceSpanFromNode(node),
+                    });
 
-                _builder.Pop();
+                    _builder.Pop();
+                }
+
+                return base.VisitCSharpStatementLiteral(node);
             }
 
-            protected SourceSpan? BuildSourceSpanFromNode(SyntaxTreeNode node)
+            protected SourceSpan? BuildSourceSpanFromNode(SyntaxNode node)
             {
-                if (node == null || node.Start == SourceLocation.Undefined)
+                if (node == null)
                 {
                     return null;
                 }
 
-                var span = new SourceSpan(
-                    node.Start.FilePath ?? FilePath,
-                    node.Start.AbsoluteIndex,
-                    node.Start.LineIndex,
-                    node.Start.CharacterIndex,
-                    node.Length);
-                return span;
+                if (node.Position == SourceDocument.Length)
+                {
+                    // This can happen when we have a marker symbol at the end of the syntax tree.
```

Comment:
This block isn't needed anymore. I've handled this in GetSourceSpan. I'll remove.